File: java/src/main/java/it/giannibombelli/wsc2026/payment/application/services/PaymentProcessing.java
package it.giannibombelli.wsc2026.payment.application.services;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.common.application.UseCase;
import it.giannibombelli.wsc2026.common.application.events.EventPublisher;
import it.giannibombelli.wsc2026.common.domain.identity.EntityId;
import it.giannibombelli.wsc2026.payment.application.commands.StartTransaction;
import it.giannibombelli.wsc2026.payment.domain.events.PaymentEvent;
import it.giannibombelli.wsc2026.payment.domain.events.TransactionAccepted;
import it.giannibombelli.wsc2026.payment.domain.events.TransactionRejected;
import it.giannibombelli.wsc2026.payment.domain.events.TransactionStarted;
import it.giannibombelli.wsc2026.payment.domain.payment.Payment;
import it.giannibombelli.wsc2026.payment.domain.payment.Provider;
import it.giannibombelli.wsc2026.payment.domain.payment.TransactionId;
import it.giannibombelli.wsc2026.payment.domain.policies.PaymentCharging;
import it.giannibombelli.wsc2026.payment.domain.ports.PaymentProvider;
import it.giannibombelli.wsc2026.payment.domain.ports.PaymentProviderResult;
import it.giannibombelli.wsc2026.payment.domain.ports.PaymentRepository;

import java.util.Map;

import static java.util.Objects.requireNonNull;

public final class PaymentProcessing implements UseCase<StartTransaction, TransactionStarted> {
    private final PaymentRepository paymentRepository;
    private final Map<String, PaymentProvider> providers;
    private final EventPublisher<PaymentEvent> eventPublisher;

    public PaymentProcessing(PaymentRepository paymentRepository,
                             Map<String, PaymentProvider> providers,
                             EventPublisher<PaymentEvent> eventPublisher) {
        this.paymentRepository = requireNonNull(paymentRepository);
        this.providers = requireNonNull(providers);
        this.eventPublisher = requireNonNull(eventPublisher);
    }

    @Override
    public TransactionStarted invoke(StartTransaction cmd) {
        Require.requireArgument(cmd, "command");

        Payment payment = paymentRepository.findById(cmd.aggregateId())
            .orElseThrow(() -> new IllegalArgumentException("payment not found"));

        PaymentProvider paymentProvider = providers.get(cmd.provider().label());
        if (paymentProvider == null) {
            throw new IllegalArgumentException("unknown provider: " + cmd.provider().label());
        }

        TransactionId transactionId = EntityId.generate(TransactionId::new);
        TransactionStarted started = payment.startTransaction(
            transactionId,
            cmd.provider(),
            cmd.providerReference(),
            cmd.amount(),
            cmd.startedAt()
        );
        paymentRepository.save(payment);
        eventPublisher.publish(started);

        PaymentCharging charging = new PaymentCharging(paymentProvider);
        PaymentProviderResult result = charging.charge(started);

        switch (result) {
            case PaymentProviderResult.Success s -> eventPublisher.publish(new TransactionAccepted(
                payment.id(), cmd.provider(), started.transactionId(), cmd.amount(), s.providerCompletedAt()));
            case PaymentProviderResult.Failure f -> eventPublisher.publish(new TransactionRejected(
                payment.id(), cmd.provider(), started.transactionId(), f.reason()));
        }

        return started;
    }
}

File: typescript/src/payment/application/services/paymentProcessing.ts
// Servizio applicativo che carica un Payment, avvia una transazione,
// invoca il provider e pubblica gli eventi di esito.

import { generateId } from '@/common/domain/identity/entityId.js';
import { Uuid } from '@/common/domain/primitive/uuid.js';
import type { UseCase } from '@/common/application/usecase.js';
import type { EventPublisher } from '@/common/application/events/eventPublisher.js';
import { StartTransaction } from '../commands/startTransaction.js';
import { PaymentEvent } from '../../domain/events/paymentEvent.js';
import { transactionAccepted } from '../../domain/events/transactionAccepted.js';
import { transactionRejected } from '../../domain/events/transactionRejected.js';
import type { TransactionStarted } from '../../domain/events/transactionStarted.js';
import { PaymentCharging } from '../../domain/policies/paymentCharging.js';
import { PaymentNotFoundException } from '../../domain/payment/paymentNotFoundException.js';
import { TransactionId } from '../../domain/payment/transactionId.js';
import type { PaymentProvider } from '../../domain/ports/paymentProvider.js';
import { PaymentProviderResult } from '../../domain/ports/paymentProviderResult.js';
import type { PaymentRepository } from '../../domain/ports/paymentRepository.js';
import { requireArgument } from '@/common/utils/requireArgument.js';
import { requireDependency } from '@/common/utils/requireDependency.js';

export class PaymentProcessing implements UseCase<StartTransaction, TransactionStarted> {
  private readonly _paymentRepository: PaymentRepository;
  private readonly _providers: Record<string, PaymentProvider>;
  private readonly _eventPublisher: EventPublisher<PaymentEvent>;

  constructor(
    paymentRepository: PaymentRepository,
    providers: Record<string, PaymentProvider>,
    eventPublisher: EventPublisher<PaymentEvent>
  ) {
    requireDependency(paymentRepository, "paymentRepository");
    requireDependency(providers, "providers");
    requireDependency(eventPublisher, "eventPublisher");
    this._paymentRepository = paymentRepository;
    this._providers = providers;
    this._eventPublisher = eventPublisher;
  }

  invoke(cmd: StartTransaction): TransactionStarted {
    requireArgument(cmd, 'command');

    const payment = this._paymentRepository.findById(cmd.aggregateId);
    if (payment === null) throw new PaymentNotFoundException();

    const provider = this._providers[cmd.provider];
    if (provider === undefined) {
      throw new Error(`unknown provider: ${cmd.provider}`);
    }

    if (!cmd.amount.equals(payment.amount())) {
      throw new Error('payment amount must be covered by a single transaction');
    }

    const transactionId = generateId((value: Uuid) => new TransactionId(value));
    const started = payment.startTransaction(
      transactionId,
      cmd.provider,
      cmd.providerReference,
      cmd.amount,
      cmd.startedAt
    );
    this._paymentRepository.save(payment);
    this._eventPublisher.publish(started);

    const charging = new PaymentCharging(provider);
    const result: PaymentProviderResult = charging.charge(started);

    if (result.kind === 'success') {
      this._eventPublisher.publish(
        transactionAccepted(
          payment.id(),
          cmd.provider,
          started.transactionId,
          cmd.amount,
          result.providerCompletedAt
        )
      );
    } else {
      this._eventPublisher.publish(
        transactionRejected(payment.id(), cmd.provider, started.transactionId, result.reason)
      );
    }

    return started;
  }
}
